### DISCUSSION TITLE
`notcurses-demo yield` especially slow compared to kitty.

### DISCUSSION THREAD

**Discussion by ginnyTheCat** [ghostty-org/ghostty#2827] [Feature Requests, Ideas] [answered]
Running the `notcurses-demo` in ghostty and kitty gives pretty similar results. However the yield test is particularly slow.

On the left is ghostty, on the right kitty.
![screenshot of the notcurses-demo result timings](https://github.com/user-attachments/assets/75e670c1-6322-4c9f-b9d8-9d47dd72bfd5)

We tried using perf to profile shows most of the time being spent in the `Parser.next` function, and there something vectorized. I don't know well zig translates these to the source lines shown so I'm not sure how much this helps.
<details>

![Image](https://github.com/user-attachments/assets/78d01fd4-1dbe-44a4-bca8-dab17541ef6c)
![Image](https://github.com/user-attachments/assets/93d9e461-2db5-43ed-96fa-9ecbc492bc9a)
![Image](https://github.com/user-attachments/assets/0a2ff013-4eac-493a-a681-7da2068727cd)

</details>

**Comment by mitchellh** [comment]
I'd love to dig into this. The images look the same to me (or there is only one) so I can't see the comparison. 

The time being spent in `Parser.next` makes sense to me and doesn't necessarily cause me any concern. Any VT heavy stream _should_ be spending most of its time there. That's a highly optimized function with heavy vectorization that we've shown to be very fast compared to other terminals, so I suspect there is a bottleneck happening somewhere else, but more research needs to be done.

I converted to a discussion until something actionable can come out of this.

**Reply by ginnyTheCat** [reply]
The difference lies in the `yield` timings (23.4s on ghostty, 3.9s on kitty, 5.6s on wezterm). To see it running you can run `notcurses-demo yield` in your command line.

> That's a highly optimized function with heavy vectorization that we've shown to be very fast compared to other terminals

That's what i thought. Makes it more annoying to find though :sweat_smile:

**Reply by mitchellh** [reply]
Yeah, I find Linux tooling to find this stuff is unfortunately terrible compared to macOS Instruments so I'll check in there.

**Reply by mitchellh** [reply, marked as answer]
Ah, just ran the `yield` demo. This uses Kitty graphics. Kitty graphics protocol is a known area where there is a ton of optimization opportunity left and a known slow place. If anyone wants to optimize that part of the stack I'd love it. 😄

**Reply by ginnyTheCat** [reply]
The main difference I think is ghostty not supporting kitty animations. This causes `notcurses` to [not zlib compress the image data it's sending](https://github.com/dankamongmen/notcurses/blob/a5c7d2262a333353bd5c3428c9397de4864c79ff/src/lib/kitty.c#L925-L926) as well. That's prolly the reason as well why at the bottom of the results table it shows "2.3 Gi" transferred on the ghostty side while only "999Mi" on the kitty side.

The captured output with ghostty on the left and kitty on the right.
![hexdump comparing ghostty and kittys notcurses-demo output](https://github.com/user-attachments/assets/e863b2e5-fb84-4690-8d2f-ab7185f294ac)
